### ISSUE TITLE
Am I setting `timeout` correctly?

### ISSUE THREAD

**Issue by arschrier** [Studiosity/grover#143]
Hello!

I am trying to generate very large PDFs, but Grover always seems to time out after 30 seconds. Setting `timeout: 0` does not seem to solve the issue.

Creating the PDF:
```ruby
Grover.new(html_with_absolute_urls, grover_options).to_pdf
```

Grover options:
```ruby
{
  format: 'Letter',
  margin: {
    top: '0.5in',
    bottom: '0.5in',
    left: '0.5in',
    right: '0.5in'
  },
  prefer_css_page_size: true,
  emulate_media: 'screen',
  cache: false,
  timeout: 0,
  launch_args: ['--font-render-hinting=medium'],
  wait_until: 'networkidle0'
}
```

Error message:
```
Grover::JavaScript::TimeoutError: waiting for Page.printToPDF failed: timeout 30000ms exceeded
```

Grover version: 1.0.6
Puppeteer version: 10.4.0

**Comment by arschrier** [comment]
Looks like downgrading Puppeteer to 10.2.0 solved the issue.

Possibly related: https://github.com/puppeteer/puppeteer/pull/7508

**Comment by abrom** [comment]
Interesting find. Seems likely that it's related to that Puppeteer change. The Puppeteer `helper.waitWithTimeout` doesn't appear to support `0` as a way of disabling the timeout, thus it would need to be implemented in the `createPDFStream` method instead. 

Maybe:
```
let result;
if (timeout) {
  result = await helper.waitWithTimeout(
    printCommandPromise,
    'Page.printToPDF',
    timeout
  );
} else {
  result = await printCommandPromise;
}
```
in place of:
https://github.com/puppeteer/puppeteer/pull/7508/files#diff-7b571f1b17991b899111f41e9844844f2fd527b4c1a775aaeb713ceb8b17959aR2794-R2798

Although best raise this with the Puppeteer team directly

**Comment by abrom** [comment]
Although I would note that the timeout configured for Grover is only used in the request phase (ie the call to `page.goto`). See https://github.com/Studiosity/grover/blob/main/lib/grover/js/processor.js#L62-L67

The issue you referenced would seem to only apply if the page was quick-ish to fetch, but slow to convert to PDF.

**Comment by arschrier** [comment]
Thanks for the quick reply! That's what I was wondering. Most of the PDFs I generate seem to fall into that fast to request slow to convert category.

Aside from disabling the time out, would it be reasonable to pass an additional conversion `timeout` to Puppeteer?

**Comment by abrom** [comment]
Yes, definitely seems reasonable. Although it may be a good idea to have grover be able to take `request_timeout` and `convert_timeout` params that would overload any `timeout` property set (as appropriate).

**Comment by abrom** [comment]
FYI I've created #145 to add those new timeout options (as well as apply `timeout` as a fallback). Just awaiting TravidCI to top up the build credits before I merge

**Comment by abrom** [comment]
These changes have been released in [v1.1.0](https://rubygems.org/gems/grover/versions/1.1.0)

**Comment by arschrier** [comment]
Wow that was fast! Thank you!
